determined-ai/determined
# feat: Trial and Task log filters [WEB-239] [WEB-240]

## Description

Trial and task logs can have an agent, container, rank, etc. which we can use to filter these logs. When a log entry doesn't have a rank, the **current** version sees a NULL and filters it out on the server.  The returned agent/container/rank values are used to populate log filters.

In this update:
- the null value is included in API responses as an empty string.
- rankIds accepts and array of strings (to include empty string) instead of ints
- "No Container" works by sending an empty string which is the default for these logs. rankId has a number or null value so the empty string cannot be used. Query with the new `FilterOperationInOrNull`

## Test Plan

- Launch a Jupyter notebook locally and copy its task ID (such as 926e5c59-afd8-415d-b3a4-89846d194dee).
- Visit `http://localhost:8080/api/v1/tasks/:task_id/logs/fields` - confirm each array has at least one value, and it is an empty string, no NULLs, no empty values, no ints for rank.
- Run an experiment and copy its trial ID (an int, **not** the experiment ID)
- Visit `http://localhost:8080/api/v1/trials/:trial_id/logs/fields` - confirm each array has at least one value, and it is an empty string, no NULLs, no empty values, no ints for rank.
- For dropdowns: on a trial logs page, you need more than one option for a dropdown to appear. You probably will see an Agents dropdown with No Agent.
- For trial logs, you can check rankId is working by going to a trial log endpoint with rankIds set to a blank value, such as `http://localhost:8080/api/v1/trials/:trial_id/logs?rankIds=`

## Checklist
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
